0.7.1 tests failing in ppc64le with g++14 -O2

TL;DR full symptoms, gathered below:
 * test failures with g++14 -O2 cross-compiled to ppc64le
 * no failures with g++14 -O1 cross-compiled to ppc64le
 * no failures with g++14 -O3 cross-compiled to ppc64le
 * no failures with g++13 -O2 cross-compiled to ppc64le
 * no failures with g++14 -O2 on x64

### Original report
v0.7.1 is showing quite a few regressions on `ppc64le` only:

```
The following tests FAILED:
	  1 - ryml-test-quickstart (Failed)
	  7 - ryml-test-parse_engine_6_qmrk (Failed)
	  8 - ryml-test-parse_engine_7_seqimap (SEGFAULT)
	 16 - ryml-test-tree (SEGFAULT)
	 18 - ryml-test-emit (Failed)
	 19 - ryml-test-style (Failed)
	 22 - ryml-test-json (Failed)
	 24 - ryml-test-merge (Failed)
	 28 - ryml-test-seq (SEGFAULT)
	 31 - ryml-test-map (SEGFAULT)
	 38 - ryml-test-scalar_null (Failed)
	 53 - ryml-test-anchor (Failed)
	 55 - ryml-test-number (Failed)
```

[builder-live.log](https://github.com/user-attachments/files/16640818/builder-live.log)

(I am not sure if I tested 0.7.0 thoroughly on non-`x86_64` architectures or not.)

_Originally posted by @musicinmybrain in https://github.com/biojppm/rapidyaml/issues/440#issuecomment-2294036666_

So I started by trying to imitate Fedora’s build flags as closely as I reasonably could, with the idea that if I could reproduce this with `samples/add_subdirectory`, I would 

```
$ mock -r fedora-rawhide-ppc64le --clean
$ mock -r fedora-rawhide-ppc64le -i git-core cmake gcc-c++ 'cmake(gtest)'
$ mock -r fedora-rawhide-ppc64le --shell --enable-network
<mock-chroot> sh-5.2# git clone https://github.com/biojppm/rapidyaml.git
<mock-chroot> sh-5.2# cd rapidyaml/
<mock-chroot> sh-5.2# git checkout fix/464_ub_ppc64le
<mock-chroot> sh-5.2# git submodule update --init --recursive
<mock-chroot> sh-5.2# cd samples/add_subdirectory
<mock-chroot> sh-5.2# CXXFLAGS="-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection" cmake -S . -B build -DRYML_CXX_STANDARD=14
<mock-chroot> sh-5.2# cmake --build build -j16 --target run
[ 95%] Linking CXX executable ryml-quickstart
/builddir/rapidyaml/ext/c4core/src/c4/charconv.hpp:197:3: warning: type ‘c4::RealFormat_e’ violates the C++ One Definition Rule [-Wodr]
  197 | } RealFormat_e;
      |   ^
/builddir/rapidyaml/ext/c4core/src/c4/charconv.hpp:209:3: note: an enum with different value name is defined in another translation unit
  209 | } RealFormat_e;
      |   ^
/builddir/rapidyaml/ext/c4core/src/c4/charconv.hpp:190:5: note: name ‘FTOA_FLOAT’ is defined as 32-bit while another translation unit defines it as 8-bit
  190 |     FTOA_FLOAT = static_cast<std::underlying_type<std::chars_format>::type>(std::chars_format::fixed),
      |     ^
/builddir/rapidyaml/ext/c4core/src/c4/charconv.hpp:202:5: note: mismatching definition
  202 |     FTOA_FLOAT = 'f',
      |     ^
[ 95%] Built target ryml-quickstart
[100%] running: /builddir/rapidyaml/samples/add_subdirectory/build/ryml-quickstart
WTF1:
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
gmake[3]: *** [CMakeFiles/run.dir/build.make:71: CMakeFiles/run] Segmentation fault (core dumped)
```

I haven’t had a chance to try to minimize the compiler flags yet, but that looked interesting enough to go ahead and report it!

Made some progress: was able to reproduce with a gcc14.2 cross-compiler, and unable to reproduce with.

I did so by getting a docker image with a gcc14(ubuntu24.10) and gcc13(ubuntu24.04) cross-compiler:

```bash
img=ubuntu:24.04 # uses gcc13.2
img=ubuntu:24.10 # uses gcc14.2
docker pull $img ; docker run --rm -it --network host -v ~/proj:/proj $img
```

... then inside the container:
```bash
apt-get update ; apt install -y g{++,cc}-powerpc64le-linux-gnu build-essential cmake qemu-user-static
# define a function to try with different compiler flags
function try464()
{
   cxxflags="$1"
   ( set -xe ; \
   cd /proj/rapidyaml/samples/add_subdirectory/ ; \
   sys=powerpc64le ; \
   bdir=build/$sys-docker ; \
   mkdir -p $bdir ; \
   cmake -S . -B $bdir -D CMAKE_BUILD_TYPE=Release \
         -DCMAKE_CXX_FLAGS_RELEASE:STRING="$cxxflags" \
         --toolchain ../../.github/toolchains/linux-gnu-$sys.cmake  ;  \
   cmake --build $bdir --target run -j --verbose )
}
```

... and finally:
```bash
$ try464 "-O1" # succeeds
$ try464 "-O3" # succeeds
$ try464 "-O2" # fails
```

FWIW, here's the error. Notice it's missing the keys:
```
[100%] running: /proj/rapidyaml/samples/add_subdirectory/build/powerpc64le-docker/ryml-quickstart
qemu-ppc64le-static /proj/rapidyaml/samples/add_subdirectory/build/powerpc64le-docker/ryml-quickstart
WTF1:
{"": "a deer, a female deer","": "a drop of golden sun","": "a name, I call myself","": "a long long way to go"}
/proj/rapidyaml/samples/quickstart.cpp:4293: ERROR: ryml::emitrs_json<std::string>(tree) == R"({"doe": "a deer, a female deer","ray": "a drop of golden sun","me": "a name, I call myself","far": "a long long way to go"})"
/proj/rapidyaml/samples/quickstart.cpp:4294: OK! ryml::emitrs_json<std::string>(json_tree) == R"({"doe": "a deer, a female deer","ray": "a drop of golden sun","me": "a name, I call myself","far": "a long long way to go"})"
WTF2:
{"": "a deer, a female deer","": "a drop of golden sun","": "a name, I call myself","": "a long long way to go"}
/proj/rapidyaml/samples/quickstart.cpp:4299: ERROR: ss.str() == R"({"doe": "a deer, a female deer","ray": "a drop of golden sun","me": "a name, I call myself","far": "a long long way to go"})"
Completed 3 checks.
ERROR: 2/3 checks failed.
make[3]: *** [CMakeFiles/run.dir/build.make:71: CMakeFiles/run] Error 2
```

--- a/.github/toolchains/linux-gnu-powerpc64le.cmake
+++ b/.github/toolchains/linux-gnu-powerpc64le.cmake
@@ -1,0 +1,26 @@
+set(CMAKE_SYSTEM_NAME Linux)
+set(CROSS_COMPILER_PREFIX /usr)
+set(CROSS_COMPILER_TARGET powerpc64le-linux-gnu)
+set(CMAKE_SYSTEM_PROCESSOR powerpc64le)
+set(CMAKE_CROSSCOMPILING_EMULATOR qemu-ppc64le-static)
+set(CMAKE_C_COMPILER ${CROSS_COMPILER_PREFIX}/bin/${CROSS_COMPILER_TARGET}-gcc)
+set(CMAKE_CXX_COMPILER ${CROSS_COMPILER_PREFIX}/bin/${CROSS_COMPILER_TARGET}-g++)
+set(CMAKE_AR ${CROSS_COMPILER_PREFIX}/bin/${CROSS_COMPILER_TARGET}-ar CACHE FILEPATH "Archiver")
+set(CMAKE_MAKE_PROGRAM ${CROSS_COMPILER_PREFIX}/bin/make CACHE FILEPATH "make")
+
+# use -static to avoid having to install the dynamic loader
+# see https://mfo.dev.br/2018/03/17/ppc64le-on-x86_64-cross-compilers.html
+set(CMAKE_CXX_FLAGS -static)
+set(CMAKE_C_FLAGS -static)
+
+set(CMAKE_FIND_ROOT_PATH ${CROSS_COMPILER_PREFIX}/${CROSS_COMPILER_TARGET})
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+
+# needed ubuntu packages:
+#
+# sudo apt-get install -y \
+#    gcc-powerpc64le-linux-gnu \
+#    g++-powerpc64le-linux-gnu \
+#    qemu-user-system

--- a/changelog/current.md
+++ b/changelog/current.md
@@ -1,0 +1,8 @@
+## Fixes
+
+- Fix [#464](https://github.com/biojppm/rapidyaml/issues/464): test failures with g++14 -O2 in ppc64le ([PR#467](https://github.com/biojppm/rapidyaml/pull/467))
+
+
+## Thanks
+
+- @musicinmybrain

--- a/src/c4/yml/parse_engine.def.hpp
+++ b/src/c4/yml/parse_engine.def.hpp
@@ -3511,12 +3511,11 @@ csubstr ParseEngine<EventHandler>::_filter_scalar_folded(substr s, size_t indent
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_plain(ScannedScalar const& C4_RESTRICT sc, size_t indentation)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_plain(sc.scalar, indentation);
+            return _filter_scalar_plain(sc.scalar, indentation);
         }
         else
         {
@@ -3528,18 +3527,17 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_plain(ScannedScalar 
     {
         _c4dbgp("plain scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_plain(ScannedScalar const& C4_RESTRICT sc, size_t indentation)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_plain(sc.scalar, indentation);
+            return _filter_scalar_plain(sc.scalar, indentation);
         }
         else
         {
@@ -3551,7 +3549,7 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_plain(ScannedScalar 
     {
         _c4dbgp("plain scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 
@@ -3560,12 +3558,11 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_plain(ScannedScalar 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_squot(ScannedScalar const& C4_RESTRICT sc)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_squot(sc.scalar);
+            return _filter_scalar_squot(sc.scalar);
         }
         else
         {
@@ -3577,18 +3574,17 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_squot(ScannedScalar 
     {
         _c4dbgp("squo key scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_squot(ScannedScalar const& C4_RESTRICT sc)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_squot(sc.scalar);
+            return _filter_scalar_squot(sc.scalar);
         }
         else
         {
@@ -3600,7 +3596,7 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_squot(ScannedScalar 
     {
         _c4dbgp("squo val scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 
@@ -3609,12 +3605,11 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_squot(ScannedScalar 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_dquot(ScannedScalar const& C4_RESTRICT sc)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_dquot(sc.scalar);
+            return _filter_scalar_dquot(sc.scalar);
         }
         else
         {
@@ -3626,18 +3621,17 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_dquot(ScannedScalar 
     {
         _c4dbgp("dquo scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_dquot(ScannedScalar const& C4_RESTRICT sc)
 {
-    csubstr maybe_filtered = sc.scalar;
     if(sc.needs_filter)
     {
         if(m_options.scalar_filtering())
         {
-            maybe_filtered = _filter_scalar_dquot(sc.scalar);
+            return _filter_scalar_dquot(sc.scalar);
         }
         else
         {
@@ -3649,7 +3643,7 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_dquot(ScannedScalar 
     {
         _c4dbgp("dquo scalar doesn't need filtering");
     }
-    return maybe_filtered;
+    return sc.scalar;
 }
 
 
@@ -3658,33 +3652,31 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_dquot(ScannedScalar 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_literal(ScannedBlock const& C4_RESTRICT sb)
 {
-    csubstr maybe_filtered = sb.scalar;
     if(m_options.scalar_filtering())
     {
-        maybe_filtered = _filter_scalar_literal(sb.scalar, sb.indentation, sb.chomp);
+        return _filter_scalar_literal(sb.scalar, sb.indentation, sb.chomp);
     }
     else
     {
         _c4dbgp("literal scalar left unfiltered");
         m_evt_handler->mark_key_scalar_unfiltered();
     }
-    return maybe_filtered;
+    return sb.scalar;
 }
 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_literal(ScannedBlock const& C4_RESTRICT sb)
 {
-    csubstr maybe_filtered = sb.scalar;
     if(m_options.scalar_filtering())
     {
-        maybe_filtered = _filter_scalar_literal(sb.scalar, sb.indentation, sb.chomp);
+        return _filter_scalar_literal(sb.scalar, sb.indentation, sb.chomp);
     }
     else
     {
         _c4dbgp("literal scalar left unfiltered");
         m_evt_handler->mark_val_scalar_unfiltered();
     }
-    return maybe_filtered;
+    return sb.scalar;
 }
 
 
@@ -3693,33 +3685,31 @@ csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_literal(ScannedBlock
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_key_scalar_folded(ScannedBlock const& C4_RESTRICT sb)
 {
-    csubstr maybe_filtered = sb.scalar;
     if(m_options.scalar_filtering())
     {
-        maybe_filtered = _filter_scalar_folded(sb.scalar, sb.indentation, sb.chomp);
+        return _filter_scalar_folded(sb.scalar, sb.indentation, sb.chomp);
     }
     else
     {
         _c4dbgp("folded scalar left unfiltered");
         m_evt_handler->mark_key_scalar_unfiltered();
     }
-    return maybe_filtered;
+    return sb.scalar;
 }
 
 template<class EventHandler>
 csubstr ParseEngine<EventHandler>::_maybe_filter_val_scalar_folded(ScannedBlock const& C4_RESTRICT sb)
 {
-    csubstr maybe_filtered = sb.scalar;
     if(m_options.scalar_filtering())
     {
-        maybe_filtered = _filter_scalar_folded(sb.scalar, sb.indentation, sb.chomp);
+        return _filter_scalar_folded(sb.scalar, sb.indentation, sb.chomp);
     }
     else
     {
         _c4dbgp("folded scalar left unfiltered");
         m_evt_handler->mark_val_scalar_unfiltered();
     }
-    return maybe_filtered;
+    return sb.scalar;
 }
 
 

